File: 747-MinCostClimbingStairs/747-MinCostClimbingStairs.cpp
// Last updated: 02/08/2025, 18:48:43
class Solution {
    int helper(vector<int>& cost, vector<int>& dp, int n){
        if(n ==0 || n == 1){
            dp[n] = 0;
            return 0;
        }
        if(dp[n] != -1){
            return dp[n];
        }
        // int right = INT_MAX;
        int left = helper(cost,dp,n-1) + cost[n-1];
        // if(n >1){
        int right = helper(cost, dp,n-2) + cost[n-2];
        // }
        dp[n] = min(left, right);
        return dp[n];
    }
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int> dp(n+1,-1);
        helper(cost,dp,n);
        return dp[n];

    }
};

File: 747-MinCostClimbingStairs/747-MinCostClimbingStairs.py
# Last updated: 01/10/2025, 00:12:35
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n = len(cost)
        if n == 1:
            return cost[0]
        if n == 2:
            return min(cost[1],cost[0])
        prev1 = cost[0]
        prev2 = cost[1]
        for i in range(2,n):
            current = cost[i] + min(prev1,prev2)
            prev1 = prev2
            prev2 = current
        return min(prev1,prev2)